trinity/plugins/builtin/peer_discovery/plugin.py

```diff
@@ -0,0 +1,151 @@
+from argparse import (
+    ArgumentParser,
+    _SubParsersAction,
+)
+import asyncio
+
+from eth_typing import (
+    BlockNumber,
+)
+from lahja import (
+    Endpoint,
+)
+
+from eth.constants import (
+    GENESIS_BLOCK_NUMBER
+)
+
+from p2p.discovery import (
+    get_v5_topic,
+    DiscoveryByTopicProtocol,
+    PreferredNodeDiscoveryProtocol,
+)
+from p2p.kademlia import (
+    Address,
+)
+from p2p.protocol import (
+    Protocol,
+)
+from p2p.service import (
+    BaseService,
+)
+
+from trinity.config import (
+    TrinityBeaconConfig,
+    TrinityConfig,
+)
+from trinity.db.manager import (
+    create_db_manager
+)
+from trinity.extensibility import (
+    BaseIsolatedPlugin,
+)
+from trinity.protocol.bcc.proto import (
+    BCCProtocol,
+)
+from trinity.protocol.eth.proto import (
+    ETHProtocol,
+)
+from trinity.protocol.les.proto import (
+    LESProtocolV2,
+)
+from trinity.utils.shutdown import (
+    exit_with_service_and_endpoint,
+)
+from .discovery_service import (
+    DiscoveryService,
+)
+
+
+def get_protocol(trinity_config: TrinityConfig) -> Protocol:
+    # For now DiscoveryByTopicProtocol supports a single topic, so we use the latest
+    # version of our supported protocols. Maybe this could be more generic?
+    if isinstance(trinity_config, TrinityBeaconConfig):
+        return BCCProtocol
+    elif trinity_config.is_light_mode:
+        return LESProtocolV2
+    else:
+        return ETHProtocol
+
+
+def get_discv5_topic(trinity_config: TrinityConfig, protocol: Protocol):
+    db_manager = create_db_manager(trinity_config.database_ipc_path)
+    db_manager.connect()
+
+    header_db = db_manager.get_headerdb()
+    genesis_hash = header_db.get_canonical_block_hash(BlockNumber(GENESIS_BLOCK_NUMBER))
+
+    return get_v5_topic(protocol, genesis_hash)
+
+
+class DiscoveryBootstrapService(BaseService):
+    """
+    Bootstrap discovery to provide a parent ``CancellationToken``
+    """
+
+    def __init__(self, event_bus: Endpoint, trinity_config: TrinityConfig) -> None:
+        super().__init__()
+        self.event_bus = event_bus
+        self.trinity_config = trinity_config
+
+    async def _run(self) -> None:
+        external_ip = "0.0.0.0"
```

Comment:
@pipermerriam in current `master` we use the (experimental) upnpservice to *maybe* resolve a different external IP first. I'm honestly not so sure how important that is or if it ever worked. I'm pretty sure I never had this working locally here but I guess it just doesn't apply to the local home network that I usually work from.

Reply:
I believe it does indeed work, it just isn't highly reliable and when it fails it is a little noisy towards the user.

Reply:
But do you think it is ok to leave this hardcoded with `0.0.0.0`? The way I understand the upnpservice, it runs continuously to do port mapping so it is not just there to potentially give us a different external IP for discovery but also for later stages. Basically, I'm trying to figure out if this would be something that needs to be resolved for the discovery or if it is acceptable to go without it (and maybe create an issue to follow up on it)

Reply:
Can confirm: `upnpservice` does provide an external IP/port pair for me on a network.

Not sure about hard-coding to `0.0.0.0`. It sounds bad, but I won't be able to dig in today. So I guess... try it? :stuck_out_tongue:

Reply:
@carver do you prefer to keep this open until you have tried this on your upnp network? I also think we should move the UpnpService in an isolated plugin as well and then we can also put back that first external ip lookup. I don't think that service does much right now though. Apart from the initial lookup of the external ip, it is supposed to continuously provide a port mapping but I don't see how that would be used anywhere.

Reply:
No it's fine, I'll end up testing it in master :+1:

Reply:
> Apart from the initial lookup of the external ip, it is supposed to continuously provide a port mapping but I don't see how that would be used anywhere.

I could be mistaken, but I think the idea is that you have to keep pinging the router so that incoming requests to that port get directed to your computer. So it would be necessary for other peers to create incoming connections. Incoming peer connections is not a thing that seems to be working well (at all?) already, so definitely not worth holding up the PR.